test/lit/passes/type-merging.wast

```diff
@@ -0,0 +1,198 @@
+    (local $a (ref null $A))
+    ;; $B can be merged into $A.
+    (local $b (ref null $B))
+    ;; $C cannot because it adds a field.
+    (local $c (ref null $C))
+    ;; $D cannot because it has a cast.
+    (local $d (ref null $D))
+
+    ;; A cast of $A has no effect.
+    (drop
+      (ref.cast_static $A
+        (local.get $a)
+      )
+    )
+    ;; A cast of $D prevents it from being merged.
+    (drop
+      (ref.cast_static $D
+        (local.get $a)
+      )
+    )
+  )
+)
+
+;; Multiple levels of merging.
+(module
+  ;; CHECK:      (type $A (struct (field i32)))
+  (type $A (struct_subtype (field i32) data))
+  (type $B (struct_subtype (field i32) $A))
+  (type $C (struct_subtype (field i32) $B))
+  ;; CHECK:      (type $D (struct_subtype (field i32) (field f64) $A))
+  (type $D (struct_subtype (field i32) (field f64) $A))
+  (type $E (struct_subtype (field i32) (field f64) $D))
+  (type $F (struct_subtype (field i32) (field f64) $E))
+  (type $G (struct_subtype (field i32) (field f64) $F))
+
+  ;; CHECK:      (type $none_=>_none (func))
+
+  ;; CHECK:      (func $foo (type $none_=>_none)
+  ;; CHECK-NEXT:  (local $a (ref null $A))
+  ;; CHECK-NEXT:  (local $b (ref null $A))
+  ;; CHECK-NEXT:  (local $c (ref null $A))
+  ;; CHECK-NEXT:  (local $d (ref null $D))
+  ;; CHECK-NEXT:  (local $e (ref null $D))
+  ;; CHECK-NEXT:  (local $f (ref null $D))
+  ;; CHECK-NEXT:  (local $g (ref null $D))
+  ;; CHECK-NEXT:  (nop)
+  ;; CHECK-NEXT: )
+  (func $foo
+    (local $a (ref null $A))
+    ;; $B can be merged into $A.
+    (local $b (ref null $B))
+    ;; $C can be merged into $B, so it will merge into $A.
+    (local $c (ref null $C))
+    ;; $D cannot be merged into $A as it adds a field.
+    (local $d (ref null $D))
+    ;; $E can be merged into $D.
+    (local $e (ref null $E))
+    ;; $F can be merged into $E, so it will merge into $D.
+    (local $f (ref null $F))
+    ;; $G can be merged into $F, so it will merge into $D.
+    (local $g (ref null $G))
+  )
+)
+
+(module
+  ;; CHECK:      (type $A (struct (field (ref null $A))))
+  (type $A (struct_subtype (field (ref null $A)) data))
+  (type $B (struct_subtype (field (ref null $A)) $A))
+  ;; CHECK:      (type $none_=>_none (func))
+
+  ;; CHECK:      (type $C (struct_subtype (field (ref null $A)) $A))
+  (type $C (struct_subtype (field (ref null $B)) $A))
+
+  ;; CHECK:      (func $foo (type $none_=>_none)
+  ;; CHECK-NEXT:  (local $a (ref null $A))
+  ;; CHECK-NEXT:  (local $b (ref null $A))
+  ;; CHECK-NEXT:  (local $c (ref null $C))
+  ;; CHECK-NEXT:  (nop)
+  ;; CHECK-NEXT: )
+  (func $foo
+    ;; $A will remain the same.
+    (local $a (ref null $A))
+    ;; $B can be merged into $A.
+    (local $b (ref null $B))
+    ;; $C refines the field, so it cannot be merged. However, separately, in
+    ;; the type definition of $C, its field of type $B should become $A. That
+    ;; is, $B should no longer be used anywhere.
+    (local $c (ref null $C))
```

Comment:
Do you mean a topological sort based on the fields? I'm not sure how to reason about that in the presence of recursive types. How about a TODO? :smile:

Reply:
No, just a topo sort on the subtype relation so that parents are processed before their children. You could even use `ModuleUtils::getOptimizedIndexedHeapType`, which already does that topo sort (although it does other work as well that wouldn't be useful here.

(Processing parents before children is already built into the processing algorithm I proposed, but would work fine with the memoization algorithm as well as long as there is a separate topo sort.)

Reply:
Sorry, I don't understand. Maybe I'm not seeing what algorithm you're proposing here? Specifically, when you suggest sorting, that is to sort the order of some particular iteration, I guess - which iteration in the current algorithm do you mean? Or are you suggesting adding a new iteration?

Reply:
I mean processing parents before their children when determining whether each type should be merged. If we can guarantee that all the supertypes of a type have already been considered for merging when we consider merging that type, then we can take into account how the supertypes have been merged.

Reply:
Ok, I tried working it out and you're right that you would need to process the field types first and that can cause issues with cycles. I'm not sure why I thought processing parents first would be sufficient; sorry about that.

I'm pretty sure we can use a coinductive definition of equality here and a corecursive comparator function, just like we did with equirecursive types. A TODO sounds fine for now!